File: Cpp/剑指 Offer II/115.重建序列.cpp
//
//  115.重建序列.cpp
//  Cpp
//
//  Created by warden on 2022/7/23.
//

#include <iostream>
#include <vector>
#include <unordered_set>
#include <queue>

using namespace std;

// @lc code=start
class Solution {
public:
    bool sequenceReconstruction(vector<int>& nums, vector<vector<int>>& sequences) {
        int n = (int)nums.size();
        vector<int> indegress(n + 1);
        vector<unordered_set<int>> graph(n + 1);
        for (auto& sequence : sequences) {
            for (int i = 1; i < sequence.size(); ++i) {
                int prev = sequence[i - 1], next = sequence[i];
                if (!graph[prev].count(next)) {
                    graph[prev].emplace(next);
                    ++indegress[next];
                }
            }
        }
        queue<int> q;
        for (int i = 1; i <= n; ++i) {
            if (indegress[i] == 0) {
                q.emplace(i);
            }
        }
        while (!q.empty()) {
            if (q.size() > 1) {
                return false;
            }
            int num = q.front();
            q.pop();
            for (int next : graph[num]) {
                indegress[next]--;
                if (indegress[next] == 0) {
                    q.emplace(next);
                }
            }
        }
        return true;
    }
};
// @lc code=end

int main(int argc, const char * argv[]) {
    // insert code here...
    
    vector<int> nums;
    vector<vector<int>> sequences;
    bool res;
    
    /**
     输入：nums = [1,2,3], sequences = [[1,2],[1,3]]
     输出：false
     解释：有两种可能的超序列：[1,2,3]和[1,3,2]。
     序列 [1,2] 是[1,2,3]和[1,3,2]的子序列。
     序列 [1,3] 是[1,2,3]和[1,3,2]的子序列。
     因为 nums 不是唯一最短的超序列，所以返回false。
     */
    nums = {1,2,3};
    sequences = {{1,2},{1,3}};
    res = false;
    assert(res == Solution().sequenceReconstruction(nums, sequences));

    /**
     输入：nums = [1,2,3], sequences = [[1,2]]
     输出：false
     解释：最短可能的超序列为 [1,2]。
     序列 [1,2] 是它的子序列：[1,2]。
     因为 nums 不是最短的超序列，所以返回false。
     */
    nums = {1,2,3};
    sequences = {{1,2}};
    res = false;
    assert(res == Solution().sequenceReconstruction(nums, sequences));
    
    /**
     输入：nums = [1,2,3], sequences = [[1,2],[1,3],[2,3]]
     输出：true
     解释：最短可能的超序列为[1,2,3]。
     序列 [1,2] 是它的一个子序列：[1,2,3]。
     序列 [1,3] 是它的一个子序列：[1,2,3]。
     序列 [2,3] 是它的一个子序列：[1,2,3]。
     因为 nums 是唯一最短的超序列，所以返回true。
     */
    nums = {1,2,3};
    sequences = {{1,2},{1,3},{2,3}};
    res = true;
    assert(res == Solution().sequenceReconstruction(nums, sequences));
    
    cout << "OK~" << endl;
    
    return 0;
}

File: Swift/剑指 Offer II/115.重建序列.swift
//
//  115.重建序列.swift
//  Cpp
//
//  Created by warden on 2022/7/23.
//

import Foundation

// @lc code=start
class Solution {
    func sequenceReconstruction(_ nums: [Int], _ sequences: [[Int]]) -> Bool {
        let n = nums.count
        var indegress = [Int](repeating: 0, count: n + 1)
        var graph = [Set<Int>](repeating: [], count: n + 1)
        for sequence in sequences {
            for i in 1..<sequence.count {
                let prev = sequence[i - 1], next = sequence[i]
                if !graph[prev].contains(next) {
                    graph[prev].insert(next)
                    indegress[next] += 1
                }
            }
        }
        var q = [Int]()
        for i in 1...n {
            if indegress[i] == 0 {
                q.append(i)
            }
        }
        while !q.isEmpty {
            if q.count > 1 {
                return false
            }
            let num = q.removeFirst()
            for next in graph[num] {
                indegress[next] -= 1
                if indegress[next] == 0 {
                    q.append(next)
                }
            }
        }
        return true
    }
}
// @lc code=end

func main() {
    var nums: [Int]
    var sequences: [[Int]]
    var res: Bool

    /**
     输入：nums = [1,2,3], sequences = [[1,2],[1,3]]
     输出：false
     解释：有两种可能的超序列：[1,2,3]和[1,3,2]。
     序列 [1,2] 是[1,2,3]和[1,3,2]的子序列。
     序列 [1,3] 是[1,2,3]和[1,3,2]的子序列。
     因为 nums 不是唯一最短的超序列，所以返回false。
     */
    nums = [1,2,3]
    sequences = [[1,2],[1,3]]
    res = false
    assert(res == Solution().sequenceReconstruction(nums, sequences))

    /**
     输入：nums = [1,2,3], sequences = [[1,2]]
     输出：false
     解释：最短可能的超序列为 [1,2]。
     序列 [1,2] 是它的子序列：[1,2]。
     因为 nums 不是最短的超序列，所以返回false。
     */
    nums = [1,2,3]
    sequences = [[1,2]]
    res = false
    assert(res == Solution().sequenceReconstruction(nums, sequences))
    
    /**
     输入：nums = [1,2,3], sequences = [[1,2],[1,3],[2,3]]
     输出：true
     解释：最短可能的超序列为[1,2,3]。
     序列 [1,2] 是它的一个子序列：[1,2,3]。
     序列 [1,3] 是它的一个子序列：[1,2,3]。
     序列 [2,3] 是它的一个子序列：[1,2,3]。
     因为 nums 是唯一最短的超序列，所以返回true。
     */
    nums = [1,2,3]
    sequences = [[1,2],[1,3],[2,3]]
    res = true
    assert(res == Solution().sequenceReconstruction(nums, sequences))
}
